src/java/org/apache/cassandra/cql3/conditions/ColumnCondition.java

```diff
@@ -579,6 +775,43 @@ private static boolean mapAppliesTo(MapType<?, ?> type, Iterator<Cell<?>> iter,
             // they're equal
             return operator == Operator.EQ || operator == Operator.LTE || operator == Operator.GTE;
         }
+
+        @Override
+        protected BoundKind kind()
+        {
+            return BoundKind.MULTI_CELL_COLLECTION;
+        }
+
+        private static final BoundSerializer<MultiCellCollectionBound> serializer = new BoundSerializer<MultiCellCollectionBound>()
+        {
+            @Override
+            public void serialize(MultiCellCollectionBound bound, DataOutputPlus out, int version) throws IOException
+            {
+                serializeList(bound.serializedValues().get(), out, version, vintNullableSerializer);
+            }
+
+            @Override
+            public MultiCellCollectionBound deserialize(DataInputPlus in, int version, ColumnMetadata column, Operator operator) throws IOException
+            {
+                List<ByteBuffer> values = deserializeList(in, version, vintNullableSerializer);
+                List<Terminal> terminals;
+                if (operator.isContains() || operator.isContainsKey())
+                {
+                    terminals = values.stream().map(Constants.Value::new).collect(toList());
```

Comment:
nit/performance: Might be less garbage if this is just..

```
List<Terminal> terminals = new ArrayList<>(values.size());
...
for (ByteBuffer value : values)
    terminals.add(new Constants.Value(value));
```

Reply:
Now you make me wonder if streams can magically see that it could also preallocate. Or it the entire stream you think is on the slow side?

Reply:
If you really want to think performance... the array list is just a list of objects and if it were mutable we could just replace each element with what we want :-P

The mistake is actually that deserialize list is not a stream (and streams are not efficient in their own right).

Reply:
And really the List is itself a wasted allocation and indirection for the array and now I am just sad about Java.

Reply:
Anyways this is what I did
```
                int numTerminals = readCollectionSize(in, version);
                List<Terminal> terminals = new ArrayList<>(numTerminals);
                if (operator.isContains() || operator.isContainsKey())
                {
                    for (int i = 0; i < numTerminals; i++)
                        terminals.add(new Constants.Value(vintNullableSerializer.deserialize(in, version)));
                }
                else
                {
                    for (int i = 0; i < numTerminals; i++)
                        terminals.add(deserializeCqlCollectionAsTerm(vintNullableSerializer.deserialize(in, version), column.type, ProtocolVersion.CURRENT));
                }
```